lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/CatalogFunctionAccessControlRule.java

```diff
@@ -0,0 +1,114 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.base.security;
+
+import com.fasterxml.jackson.annotation.JsonAlias;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.trino.plugin.base.security.FunctionAccessControlRule.FunctionPrivilege;
+import io.trino.spi.connector.CatalogSchemaRoutineName;
+import io.trino.spi.function.FunctionKind;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class CatalogFunctionAccessControlRule
+{
+    public static final CatalogFunctionAccessControlRule ALLOW_ALL = new CatalogFunctionAccessControlRule(
+            Optional.empty(),
+            FunctionAccessControlRule.ALLOW_ALL);
+
+    private final Optional<Pattern> catalogRegex;
+    private final FunctionAccessControlRule functionAccessControlRule;
+
+    @JsonCreator
+    public CatalogFunctionAccessControlRule(
+            @JsonProperty("privileges") Set<FunctionPrivilege> privileges,
+            @JsonProperty("user") Optional<Pattern> userRegex,
+            @JsonProperty("role") Optional<Pattern> roleRegex,
+            @JsonProperty("group") Optional<Pattern> groupRegex,
+            @JsonProperty("catalog") Optional<Pattern> catalogRegex,
+            @JsonProperty("schema") Optional<Pattern> schemaRegex,
+            @JsonProperty("function") Optional<Pattern> tableFunctionRegex,
+            @JsonProperty("function_kinds") @JsonAlias("functionKinds") Set<FunctionKind> functionKinds)
+    {
+        this(catalogRegex, new FunctionAccessControlRule(privileges, userRegex, roleRegex, groupRegex, schemaRegex, tableFunctionRegex, functionKinds));
+    }
+
+    private CatalogFunctionAccessControlRule(Optional<Pattern> catalogRegex, FunctionAccessControlRule functionAccessControlRule)
+    {
+        this.catalogRegex = requireNonNull(catalogRegex, "catalogRegex is null");
+        this.functionAccessControlRule = requireNonNull(functionAccessControlRule, "tableFunctionAccessControlRule is null");
+        checkState(functionAccessControlRule.isTableFunction() || !functionAccessControlRule.isTableFunction() && catalogRegex.isEmpty(), "Cannot define catalog for others function kinds than TABLE");
+        if (catalogRegex.isPresent() && (functionAccessControlRule.getFunctionKinds().size() > 1 && functionAccessControlRule.isTableFunction())) {
+            throw new IllegalStateException("Cannot define catalog for others function kinds than TABLE");
+        }
```

Comment:
initially I thought this does not make sense. I've spent a lot of time to understand what is the purpose of this.
`isTableFunction` name is misleading with current implementation. 
please try to play with implementing this method as `functionKinds.equals(Set.of(TABLE))` or renaming it to `isTableFunction` -> `containsTableFunction`, and maybe adding method `doesNotContainTableFunction`, or inlining it

Reply:
Renamed `isTableFunction` into `hasTableFunctionKind`.

> please try to play with implementing this method as functionKinds.equals(Set.of(TABLE))
This not cover the case where you set `functionKinds` to `WINDOW` and `TABLE` and define `catalog` or `schema`